filebeat/input/filestream/internal/input-logfile/harvester_test.go

```diff
@@ -0,0 +1,331 @@
+	})
+
+	t.Run("assert a harvester error is handled", func(t *testing.T) {
+		mockHarvester := &mockHarvester{onRun: errorOnRun}
+		hg := testDefaultHarvesterGroup(t, mockHarvester)
+		gorountineChecker := newGoroutineChecker()
+
+		hg.Start(input.Context{Logger: logp.L(), Cancelation: context.Background()}, source)
+
+		gorountineChecker.waitOriginalCount()
+
+		_, ok := hg.readers.table[source.Name()]
+		require.False(t, ok)
+
+		err := hg.StopGroup()
+		require.Error(t, err)
+	})
+
+	t.Run("assert already locked resource has to wait", func(t *testing.T) {
+		var wg sync.WaitGroup
+		mockHarvester := &mockHarvester{onRun: correctOnRun, wg: &wg}
+		hg := testDefaultHarvesterGroup(t, mockHarvester)
+		inputCtx := input.Context{Logger: logp.L(), Cancelation: context.Background()}
+
+		r, err := lock(inputCtx, hg.store, source.Name())
+		if err != nil {
+			t.Fatalf("cannot lock source")
+		}
+
+		gorountineChecker := newGoroutineChecker()
+
+		wg.Add(1)
+		hg.Start(inputCtx, source)
+
+		locked := true
+		gorountineChecker.waitWhileMoreGoroutinesWithFunc(func() {
+			// harvester is waiting to start
+			_, ok := hg.readers.table[source.Name()]
+			if ok && locked {
+				releaseResource(r)
+				locked = false
+			}
+		})
+
+		// wait until harvester.Run is done
+		wg.Wait()
+		// wait until goroutine that started `harvester.Run` is finished
+		gorountineChecker.waitOriginalCount()
+		require.Equal(t, 1, mockHarvester.runCount)
+		require.Nil(t, hg.StopGroup())
+	})
+
+	t.Run("assert already locked resource has no problem when harvestergroup is cancelled", func(t *testing.T) {
+		mockHarvester := &mockHarvester{onRun: correctOnRun}
+		hg := testDefaultHarvesterGroup(t, mockHarvester)
+		inputCtx := input.Context{Logger: logp.L(), Cancelation: context.Background()}
+		gorountineChecker := newGoroutineChecker()
+
+		r, err := lock(inputCtx, hg.store, source.Name())
+		if err != nil {
+			t.Fatalf("cannot lock source")
+		}
+		defer releaseResource(r)
+
+		hg.Start(inputCtx, source)
+
+		gorountineChecker.waitWhileMoreGoroutinesWithFunc(func() {
+			err := hg.StopGroup()
+			require.Error(t, err)
+		})
+
+		require.Equal(t, 0, mockHarvester.runCount)
+
+		gorountineChecker.waitOriginalCount()
+	})
+}
+
+func testDefaultHarvesterGroup(t *testing.T, mockHarvester Harvester) *defaultHarvesterGroup {
+	return &defaultHarvesterGroup{
+		readers:   newReaderGroup(),
+		pipeline:  &pipelinemock.MockPipelineConnector{},
+		harvester: mockHarvester,
+		store:     testOpenStore(t, "test", nil),
+		tg:        unison.TaskGroup{},
+	}
+}
+
+type gorountineChecker struct {
+	n    int
+	wait time.Duration
+}
+
+func newGoroutineChecker() *gorountineChecker {
+	return &gorountineChecker{
+		n:    runtime.NumGoroutine(),
+		wait: 10 * time.Millisecond,
+	}
+}
+
+func (c *gorountineChecker) waitWhileMoreGoroutinesWithFunc(f func()) {
+	for c.n < runtime.NumGoroutine() {
+		fmt.Println("waiting with function", c.n, runtime.NumGoroutine())
```

Comment:
No `fmt.Printx` in tests. Better use `t.Log`. I don't think we need to print. In case we leak the tests will hang forever (test runtime prints stack trace after 10min). 

The test runner does not report a status for tests that did print to stdout/stderr, which makes our CI think at times that a test failed.